### PR TITLE
1345 Add end to end test with data comparison

### PR DESCRIPTION
**Pull request by carlosfelix2**
This PR adds a new CLI test called "compare" that reads a configuration file for test data, sends the test data to the RS API, then reads the lineage results and compares the resulting uploaded data from RS to the fake SFTP site and returns test results.  This test runs similar to the end2end test with the exception of comparing the resulting data at the end.

**LIMITATIONS:** Only supports CSV and HL7_BATCH receivers.  Tests are run sequentially.  Will open another issue to add support for HL7 single messages like for FL (did not add it in this PR to keep review smaller).

**Configuration file:**
Specifies what the input and expected reports are for the test.
  compare-test-config.csv has the following columns:
    Input File,Sender,Expected File,Organization name,Receiver name,Expected count
  Where:
    Input File = input HL7 or CSV file relative to the config file folder
    Sender = the name of the sender including organization (e.g. ignore.ignore-hl7)
    Expected File = the name of one of the expected output files relative to the config file folder
    Organization name = the name of the receiver organization (e.g. ignore)
    Receiver name = the name of the receiver (e.g. CSV)
    Expected count = The number of expected records sent as an integer
  You can specify one or more expected files for a given input file by having multiple lines with the same input
  file.  E.g.
    Input File,Sender,Expected File,Organization name,Receiver name,Expected count
    test-0001-input-covid-19.hl7,ignore.ignore-hl7,test-0001-az-covid-19-hl7.hl7,ignore,HL7_BATCH,5
    test-0001-input-covid-19.hl7,ignore.ignore-hl7,test-0001-pima-az-covid-19.csv,ignore,CSV,5

To test:
Run the following command: `./prime test --run=compare` and verify the tests run successfully.  

## Changes
- Added compare test to CLI
- Added some unit tests to make sure the comparison works as expected instead of relying on manual testing.

### Testing
- [x] Tested locally?
- [ ] Ran `quickTest all`?
- [ ] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## To Be Done
- Update the HL7 and CSV conversion tests to use a config file to support multiple schemas, plus use the CompareData functions to compare instead of the ones in their files.
- Add support for HL7 single message files to this test

### CODE REVIEW

File: prime-router/src/main/kotlin/cli/tests/CompareData.kt [hunk 1]
```diff
@@ -0,0 +1,692 @@
+package gov.cdc.prime.router.cli.tests
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/cli/tests/CompareData.kt:1]: Shouldn't we move TestReportStream down into this tests package too?  Again, sounds like a future task

File: prime-router/src/main/kotlin/cli/tests/CompareData.kt [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+                            "'$colName' - column #${j + 1}, but no expected value.  " +
+                            "Actual: '${actualRow[j].trim()}'"
+                    )
+                }
+            }
+        }
+        result.passed = result.passed and passed
+        return passed
+    }
+}
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/cli/tests/CompareData.kt:692]: Wow - you've been a busy guy!

File: prime-router/src/test/kotlin/cli/tests/CompareDataTests.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
+package gov.cdc.prime.router.cli.tests
```
Comment (jimduff-usds) [inline prime-router/src/test/kotlin/cli/tests/CompareDataTests.kt:1]: Ah, yes, testing that the test code is correct.     Excellent.

Now you need to write CompareDataTestsTests... ;)

**Review by jimduff-usds — CHANGES_REQUESTED** [review]
This is awesome!   This looks like a solid foundation on which to build expected/actual comparison tests.   Maybe we can get rid of ./quick-tests.sh, and replace with expanded tests here.   This is something I've discussed already with @SpiderSense and @san-nat 

A few minor items:

1.  Some minor suggestions in examineResponse(...), as you'll see.  Also, see below problem with fat jar run.

2. Could you create a ticket to refactor other code to use examineResponse(...).  Also to move TestReportStream.kt down into cli/tests dir  - this might be something @san-nat could take on.

2.  First time I ran it, I got this problem:

```
$ ./prime test --run compare
Running the following tests, POSTing to http://localhost:7071/api/reports:
compare             Experimental        	Send data to API based on config file, check lineage and compare data against expected
Exception in thread "main" java.io.FileNotFoundException: file:/Users/DKIQ/Documents/code/prime-data-hub/prime-router/build/libs/prime-router-0.1-SNAPSHOT-all.jar!/clitests/compare-test-files/compare-test-config.csv (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at com.github.doyaaaaaken.kotlincsv.client.CsvReader.readAllWithHeader(CsvReader.kt:67)
	at gov.cdc.prime.router.cli.tests.DataCompareTest.readTestConfig(CompareData.kt:185)
	at gov.cdc.prime.router.cli.tests.DataCompareTest.run(CompareData.kt:108)
	at gov.cdc.prime.router.cli.TestReportStream.runTests(TestReportStream.kt:208)
	at gov.cdc.prime.router.cli.TestReportStream.run(TestReportStream.kt:190)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:204)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:213)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:17)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:396)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:393)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:411)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:436)
	at gov.cdc.prime.router.cli.MainKt.main(main.kt:268)

$ git status
On branch cfelix/1345-test-data-comparison
Your branch is up to date with 'origin/cfelix/1345-test-data-comparison'.
```

So then I ran
`./gradlew primeCLI --args='test --run compare'`
which worked great!

So, fatjar is looking in the wrong place for the clitest dir.

File: prime-router/src/main/kotlin/cli/TestReportStream.kt [hunk 4]
```diff
@@ -341,6 +345,62 @@ abstract class CoolTest {
...
+        }
+        return reportId
+    }
+
+    /**
+     * Examine the [jsonResponse] from the API, makes sure there is at least one destination reported
+     * and report any errors
+     * @return true if there are no errors in the reponse, false otherwise
+     */
+    fun examineRespose(jsonResponse: String): Boolean {
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/cli/TestReportStream.kt:366]: I went through the code, but the tests tend to use different checks.  I refactored a few more that I could.

File: prime-router/src/main/kotlin/cli/TestReportStream.kt [hunk 5]
```diff
@@ -341,6 +345,62 @@ abstract class CoolTest {
...
+            if (topic != null && !topic.isNull && topic.textValue().equals("covid-19", true)) {
+                good("'topic' is in response and correctly set to 'covid-19'")
+            } else {
+                passed = bad("***$name Test FAILED***: 'topic' is missing from response json")
+            }
+
+            if (errorCount != null && !errorCount.isNull && errorCount.intValue() == 0) {
+                good("No errors detected.")
+            } else {
+                passed = bad("***$name Test FAILED***: There were errors reported.")
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/cli/TestReportStream.kt:386]: Actually, this was done on purpose.  In this one method once you get a bad then passed is false.  Note that passed is only changed when there is a bad.  No need for this change.

File: prime-router/src/main/kotlin/cli/tests/CompareData.kt (hunk 2 above)
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/cli/tests/CompareData.kt:692]: Oh yes!

File: prime-router/src/test/kotlin/cli/tests/CompareDataTests.kt (hunk 3 above)
Comment (carlosfelix2) [reply prime-router/src/test/kotlin/cli/tests/CompareDataTests.kt:1]: I know.  I had to.  I really was not feeling it doing manual testing and found bugs once I added these unit tests.

File: prime-router/src/main/kotlin/cli/tests/CompareData.kt (hunk 1 above)
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/cli/tests/CompareData.kt:1]: I can move it.  Not a problem.

**Comment by carlosfelix2** [comment]
> This is awesome! This looks like a solid foundation on which to build expected/actual comparison tests. Maybe we can get rid of ./quick-tests.sh, and replace with expanded tests here. This is something I've discussed already with @SpiderSense and @san-nat
> 
> A few minor items:
> 
> 1. Some minor suggestions in examineResponse(...), as you'll see.  Also, see below problem with fat jar run.
> 2. Could you create a ticket to refactor other code to use examineResponse(...).  Also to move TestReportStream.kt down into cli/tests dir  - this might be something @san-nat could take on.
> 3. First time I ran it, I got this problem:
> 
> ```
> $ ./prime test --run compare
> Running the following tests, POSTing to http://localhost:7071/api/reports:
> compare             Experimental        	Send data to API based on config file, check lineage and compare data against expected
> Exception in thread "main" java.io.FileNotFoundException: file:/Users/DKIQ/Documents/code/prime-data-hub/prime-router/build/libs/prime-router-0.1-SNAPSHOT-all.jar!/clitests/compare-test-files/compare-test-config.csv (No such file or directory)
> 	at java.base/java.io.FileInputStream.open0(Native Method)
> 	at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
> 	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
> 	at com.github.doyaaaaaken.kotlincsv.client.CsvReader.readAllWithHeader(CsvReader.kt:67)
> 	at gov.cdc.prime.router.cli.tests.DataCompareTest.readTestConfig(CompareData.kt:185)
> 	at gov.cdc.prime.router.cli.tests.DataCompareTest.run(CompareData.kt:108)
> 	at gov.cdc.prime.router.cli.TestReportStream.runTests(TestReportStream.kt:208)
> 	at gov.cdc.prime.router.cli.TestReportStream.run(TestReportStream.kt:190)
> 	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:204)
> 	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:213)
> 	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:17)
> 	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:396)
> 	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:393)
> 	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:411)
> 	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:436)
> 	at gov.cdc.prime.router.cli.MainKt.main(main.kt:268)
> 
> $ git status
> On branch cfelix/1345-test-data-comparison
> Your branch is up to date with 'origin/cfelix/1345-test-data-comparison'.
> ```
> 
> So then I ran
> `./gradlew primeCLI --args='test --run compare'`
> which worked great!
> 
> So, fatjar is looking in the wrong place for the clitest dir.

1. Looked and done.
2. I just did the refactor.  It was not much, so no ticket here.
3. Good one.  The problem is that when loading resources in a JAR you can only get streams and not use a path since the path will never be valid.  Changed the code and tested running the test via ./prime CLI, gradle CLI, and in IntelliJ and now works.

File: prime-router/src/main/kotlin/cli/TestReportStream.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/cli/TestReportStream.kt:395]: Ah, yes, my "bad".

**Review by jimduff-usds — APPROVED** [review]

**Pull request merged**